### PR TITLE
feat(azure):Add support for Azure Automation

### PR DESCRIPTION
**Pull request by NeverovDS**
Issue #707

### CODE REVIEW

**Comment by NeverovDS** [comment]
@ekifel ready for review

File: internal/providers/terraform/azure/automation_account.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+		ReferenceAttributes: []string{
+			"resource_group_name",
+		},
+	}
+}
+
+func NewAzureRMAutomationAccount(d *schema.ResourceData, u *schema.UsageData) *schema.Resource {
+	var monthlyJobRunMins, nonAzureConfigNodeCount, monthlyWatcherHours *decimal.Decimal
+	group := d.References("resource_group_name")[0]
+	location := strings.ToLower(group.Get("location").String())
```
Comment (ekifel) [inline internal/providers/terraform/azure/automation_account.go:25]: We have got a new func for regions, look at [this](https://github.com/infracost/infracost/blob/7724f508e36c745043fa94ad73d7ce6e5e84797b/internal/providers/terraform/azure/util.go#L17-L45), and use it for every new Azure resources, please

File: internal/providers/terraform/azure/automation_account.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+	costComponents := make([]*schema.CostComponent, 0)
+
+	if u != nil && u.Get("monthly_job_run_mins").Type != gjson.Null {
+		monthlyJobRunMins = decimalPtr(decimal.NewFromInt(u.Get("monthly_job_run_mins").Int()))
+		if monthlyJobRunMins.IsPositive() {
+			costComponents = append(costComponents, runTimeCostComponent(location, "500", "Basic Runtime", monthlyJobRunMins))
+		}
+	} else {
+		var unknown *decimal.Decimal
+		costComponents = append(costComponents, runTimeCostComponent(location, "500", "Basic Runtime", unknown))
```
Comment (ekifel) [inline internal/providers/terraform/azure/automation_account.go:35]: You can use existing variable here (and all the cases below), not unknown
```suggestion
              costComponents = append(costComponents, runTimeCostComponent(location, "500", "Basic Runtime", monthlyJobRunMins))
```

File: internal/providers/terraform/azure/automation_job_schedule.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+		ReferenceAttributes: []string{
+			"resource_group_name",
+		},
+	}
+}
+
+func NewAzureRMAutomationJobSchedule(d *schema.ResourceData, u *schema.UsageData) *schema.Resource {
+	var monthlyJobRunMins *decimal.Decimal
+	group := d.References("resource_group_name")[0]
+	location := strings.ToLower(group.Get("location").String())
```
Comment (ekifel) [inline internal/providers/terraform/azure/automation_job_schedule.go:25]: The same thing with location, use new func here, please

**Review by ekifel — APPROVED** [review]
Thanks @NeverovDS, great job! Only some minor changes left

**Comment by NeverovDS** [comment]
@ekifel ready

**Comment by ekifel** [comment]
@alikhajeh1 LGTM!

File: internal/providers/terraform/azure/automation_account.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+		RFunc: NewAzureRMAutomationAccount,
+		ReferenceAttributes: []string{
+			"resource_group_name",
+		},
+	}
+}
+
+func NewAzureRMAutomationAccount(d *schema.ResourceData, u *schema.UsageData) *schema.Resource {
+	var monthlyJobRunMins, nonAzureConfigNodeCount, monthlyWatcherHours *decimal.Decimal
+	location := lookupRegion(d, []string{"resource_group_name"})
```
Comment (alikhajeh1) [inline internal/providers/terraform/azure/automation_account.go:23]:
```suggestion
	location := lookupRegion(d, []string{})
```


This is not needed as lookupRegion automatically looks for this.

File: internal/providers/terraform/azure/automation_account.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+	"github.com/tidwall/gjson"
+)
+
+func GetAzureRMAutomationAccountRegistryItem() *schema.RegistryItem {
+	return &schema.RegistryItem{
+		Name:  "azurerm_automation_account",
+		RFunc: NewAzureRMAutomationAccount,
+		ReferenceAttributes: []string{
+			"resource_group_name",
+		},
```
Comment (alikhajeh1) [inline internal/providers/terraform/azure/automation_account.go:17]:
```suggestion
```


This is not needed as lookupRegion automatically looks for this.

File: internal/providers/terraform/azure/automation_dsc_configuration.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+	"github.com/infracost/infracost/internal/schema"
+	"github.com/shopspring/decimal"
+	"github.com/tidwall/gjson"
+)
+
+func GetAzureRMAutomationDscConfigurationRegistryItem() *schema.RegistryItem {
+	return &schema.RegistryItem{
+		Name:  "azurerm_automation_dsc_configuration",
+		RFunc: NewAzureRMAutomationDscConfiguration,
+		ReferenceAttributes: []string{
```
Comment (alikhajeh1) [inline internal/providers/terraform/azure/automation_dsc_configuration.go:15]: same comment as above applies here.

File: internal/providers/terraform/azure/automation_dsc_configuration.go [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+	return &schema.Resource{
+		Name:           d.Address,
+		CostComponents: nodesCostComponent(d, u),
+	}
+}
+
+func nodesCostComponent(d *schema.ResourceData, u *schema.UsageData) []*schema.CostComponent {
+	var nonAzureConfigNodeCount *decimal.Decimal
+
+	location := lookupRegion(d, []string{"resource_group_name"})
```
Comment (alikhajeh1) [inline internal/providers/terraform/azure/automation_dsc_configuration.go:31]: same comment as above applies here.

File: internal/providers/terraform/azure/automation_job_schedule.go [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+	"github.com/infracost/infracost/internal/schema"
+	"github.com/shopspring/decimal"
+	"github.com/tidwall/gjson"
+)
+
+func GetAzureRMAutomationJobScheduleRegistryItem() *schema.RegistryItem {
+	return &schema.RegistryItem{
+		Name:  "azurerm_automation_job_schedule",
+		RFunc: NewAzureRMAutomationJobSchedule,
+		ReferenceAttributes: []string{
```
Comment (alikhajeh1) [inline internal/providers/terraform/azure/automation_job_schedule.go:15]: same comment as above applies here.

File: internal/providers/terraform/azure/automation_job_schedule.go [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+			"resource_group_name",
+		},
+	}
+}
+
+func NewAzureRMAutomationJobSchedule(d *schema.ResourceData, u *schema.UsageData) *schema.Resource {
+	var monthlyJobRunMins *decimal.Decimal
+	location := lookupRegion(d, []string{"resource_group_name"})
```
Comment (alikhajeh1) [inline internal/providers/terraform/azure/automation_job_schedule.go:23]: same comment as above applies here.

File: internal/providers/terraform/azure/testdata/automation_job_schedule_test/automation_job_schedule_test.golden [hunk 10]
```diff
@@ -0,0 +1,15 @@
+
+ Name                                              Monthly Qty  Unit                Monthly Cost 
+                                                                                                 
+ azurerm_automation_job_schedule.fiveMinutes                                                     
+ └─ Job run time                                             5  minutes                    $0.01 
+                                                                                                 
+ azurerm_automation_job_schedule.withoutUsage                                                    
+ └─ Job run time                               Monthly cost depends on usage: $0.002 per minutes 
+                                                                                                 
+ azurerm_automation_job_schedule.zeroMinutes
```
Comment (alikhajeh1) [inline internal/providers/terraform/azure/testdata/automation_job_schedule_test/automation_job_schedule_test.golden:10]: nice test case!

**Review by alikhajeh1 — COMMENTED** [review]
Great job @NeverovDS!
Awesome review @ekifel - I added a few minor comments.

I also updated https://github.com/infracost/infracost/commit/b6578fe5879dbd04a0942d981d51bca4b5a13f71 to clarify the point from my review.

File: internal/providers/terraform/azure/automation_account.go (hunk 4 above)
Comment (NeverovDS) [reply internal/providers/terraform/azure/automation_account.go]: @alikhajeh1 Done!

File: internal/providers/terraform/azure/automation_account.go (hunk 5 above)
Comment (NeverovDS) [reply internal/providers/terraform/azure/automation_account.go]: @alikhajeh1 Done!

**Comment by alikhajeh1** [comment]
@NeverovDS can you please check these, they're from the latest master branch. It looks like some filters might need to be updated or some location lookup logic isn't quite working.

```
=== CONT  TestAzureRMAutomationDscConfiguration
time="2021-06-07T10:05:29Z" level=warning msg="Using eastus for resource azurerm_automation_dsc_nodeconfiguration.withoutUsage as its 'location' property could not be found."
time="2021-06-07T10:05:29Z" level=warning msg="Using eastus for resource azurerm_automation_dsc_nodeconfiguration.zeroNodes as its 'location' property could not be found."
time="2021-06-07T10:05:29Z" level=warning msg="Using eastus for resource azurerm_automation_dsc_nodeconfiguration.fiveNodes as its 'location' property could not be found."
time="2021-06-07T10:05:29Z" level=warning msg="Multiple products found for azurerm_automation_account.example Watchers, using the first product"
--- PASS: TestAzureRMAutomationDscNodeconfiguration (8.60s)
=== CONT  TestAzureRMAutomationAccount
time="2021-06-07T10:05:37Z" level=warning msg="Multiple products found for azurerm_automation_account.example Watchers, using the first product"
--- PASS: TestAzureRMAutomationDscConfiguration (8.66s)
=== CONT  TestAzureRMAppServiceCustomHostnameBinding
time="2021-06-07T10:05:37Z" level=warning msg="Multiple products found for azurerm_automation_account.without_usage Watchers, using the first product"
time="2021-06-07T10:05:37Z" level=warning msg="Multiple products found for azurerm_automation_account.default Watchers, using the first product"
time="2021-06-07T10:05:37Z" level=warning msg="Multiple products found for azurerm_automation_account.allUsageEx Watchers, using the first product"
time="2021-06-07T10:05:37Z" level=warning msg="Multiple products found for azurerm_automation_account.someUsageEx Watchers, using the first product"
```

**Comment by NeverovDS** [comment]
@alikhajeh1 What location did you use?

**Comment by alikhajeh1** [comment]
I just ran `make test_azure` so it's running the tests and outputting these, you can run the same locally or just run the test for these files, e.g. `INFRACOST_LOG_LEVEL=warn go test -v -cover ./internal/providers/terraform/aws/ebs_volume_test.go`

**Pull request merged**